utils/symbol-rawelf.c

```diff
@@ -13,6 +13,96 @@
 #include "utils/utils.h"
 #include "utils/symbol-rawelf.h"
 
+/*
+ *  ELF File Header validation logic.
+ */
+int elf_validation(struct uftrace_elf_data *elf)
+{
+	Elf_Ehdr *ehdr;
+	int eclass, data, version;
+	int res = 0;
+	unsigned long mapped_saddr, mapped_eaddr, mapped_size, addr;
+
+	ehdr = &elf->ehdr;
+
+	// validate ELF Magic.
+	if (memcmp(ehdr, ELFMAG, SELFMAG)) {
+		pr_dbg("ELF Signature not matched\n");
+		res = -1;
+	}
+
+	// validate some field of elf header.
+	eclass = (int) ((unsigned char *) ehdr)[EI_CLASS];
+	data = (int) ((unsigned char *) ehdr)[EI_DATA];
+	version = (int) ((unsigned char *) ehdr)[EI_VERSION];
+
+	if (!(eclass > ELFCLASSNONE && eclass < ELFCLASSNUM)) {
+		pr_dbg("eclass out of range. it must be inside range "\
+				"between %d and %d, but Yours is [%d]\n",
+				ELFCLASSNONE, ELFCLASSNUM, eclass);
+		res = -1;
+	}
+
+	if (!(data > ELFDATANONE && data < ELFDATANUM)) {
+		pr_dbg("data out of range. it must be inside range "\
+				"between %d and %d, but Yours is [%d]\n",
+				ELFDATANONE, ELFDATANUM, data);
+		res = -1;
+	}
+
+	if (!(version > EV_NONE && version < EV_NUM)) {
+		pr_dbg("version out of range. it must be inside range "\
+				"between %d and %d, but Yours is [%d]\n",
+				EV_NONE, EV_NUM, version);
+		res = -1;
+	}
+
+	// validate ELF file size.
+	mapped_saddr = (long)elf->file_map;
+	mapped_size = elf->file_size;
+	mapped_eaddr = mapped_saddr + mapped_size;
+
+	if (ehdr->e_phnum == 0 || ehdr->e_phentsize == 0) {
+		pr_dbg("Program Header of ELF Header is not normal.\n"
+				"Program Header: Num:[%d] Size:[%d]\n",
+				ehdr->e_phnum, ehdr->e_phentsize);
+		res = -1;
+	}
+
+	if (ehdr->e_shnum > 0 && ehdr->e_shentsize == 0) {
+		pr_dbg("Section Header of ELF Header size cannot be 0.\n");
+		res = -1;
+	}
+
+	// validate Section Header
+	addr = mapped_saddr + ehdr->e_shoff +
+		ehdr->e_shnum * ehdr->e_shentsize;
+
+	if (addr > mapped_eaddr) {
+		pr_dbg("Section Header address must not over the " \
+				"file size that mapped\n" \
+				"mapped address : [%llx - %llx]" \
+				"but Yours is : [%llx]\n",
+				mapped_saddr, mapped_eaddr, addr);
+		res = -1;
+	}
```

Comment:
I was not sure it'd better failing out when it cannot find section headers.  It's legal to access ELF binary using program headers only.

Reply:
I'll move the logic that validates the ELF section header to another function and change the function that refers to the section header to call this function before reference it.